integrations/opensearch/src/haystack_integrations/document_stores/opensearch/document_store.py

```diff
@@ -300,35 +303,68 @@ def _bm25_retrieval(
         :param top_k: Maximum number of Documents to return, defaults to 10
         :param scale_score: If `True` scales the Document`s scores between 0 and 1, defaults to False
         :param all_terms_must_match: If `True` all terms in `query` must be present in the Document, defaults to False
+        :param custom_query: The query containing a mandatory `$query` and an optional `$filters` placeholder
+
+            **An example custom_query:**
+
+            ```python
+            {
+                "query": {
+                    "bool": {
+                        "should": [{"multi_match": {
+                            "query": $query,                 // mandatory query placeholder
+                            "type": "most_fields",
+                            "fields": ["content", "title"]}}],
+                        "filter": $filters                  // optional filter placeholder
+                    }
+                }
+            }
+            ```
+
         :raises ValueError: If `query` is an empty string
         :returns: List of Document that match `query`
         """
 
         if not query:
-            msg = "query must be a non empty string"
-            raise ValueError(msg)
+            body: Dict[str, Any] = {"query": {"bool": {"must": {"match_all": {}}}}}
+            if filters:
+                body["query"]["bool"]["filter"] = normalize_filters(filters)
+
+        if isinstance(custom_query, str):
+            template = Template(custom_query)
+            # substitute placeholder for query and filters for the custom_query template string
+            substitutions = {
+                "query": json.dumps(query),
+                "filters": json.dumps(normalize_filters(filters)),
+            }
+            custom_query_json = template.substitute(**substitutions)
+            body = json.loads(custom_query_json)
+        elif isinstance(custom_query, dict):
+            body = self._render_custom_query(custom_query, {"$query": query, "$filters": normalize_filters(filters)})
```

Comment:
same here, `query` can be `None` - will that be OK?

Reply:
Here it's ok. We let OpenSearch handle it.